Source/NSStoryboard.m

```diff
@@ -0,0 +1,237 @@
+                  [customObject addAttribute: usrLabel];
+                  [customObject addAttribute: customCls];
+                  
+                  // Add it to the document
+                  [objects addChild: customObject];
+                  [objects detach];
+                  [doc addChild: objects];
+                  RELEASE(customObject);
+                  
+                  // Assign application scene...
+                  ASSIGN(_applicationSceneId, sceneId);
+                }
+              else
+                {
+                  [child detach];
+                  [doc addChild: child];
+                }
+              
+              // Create document...
+              document = [[NSXMLDocument alloc] initWithRootElement: doc];
+              [_scenesMap setObject: document
+                             forKey: sceneId];
+              RELEASE(document);
+            }
+        }
+    }
+  else
+    {
+      NSLog(@"No document element in storyboard file");
+    }
+
+  NSLog(@"map = %@", _scenesMap);
+  NSLog(@"initial = %@", _initialViewControllerId);
+}
+
+// Private instance methods...
+- (id) initWithName: (NSStoryboardName)name
+             bundle: (NSBundle *)bundle
+{
+  self = [super init];
+  if (self != nil)
+    {
+      NSString *path = [bundle pathForResource: name
+                                        ofType: @"storyboard"];
+      NSData *data = [NSData dataWithContentsOfFile: path];
+      
+      NSXMLDocument *storyboardXml = [[NSXMLDocument alloc] initWithData: data
+                                                                 options: 0
+                                                                   error: NULL];
+      AUTORELEASE(storyboardXml);
+
+      [self _processStoryboard: storyboardXml];
+    }
+  return self;
+}
+
+// Class methods...
++ (void) setMainStoryboard: (NSStoryboard *)storyboard  // private, only called from NSApplicationMain()
+{
+  mainStoryboard = storyboard;
+}
+
++ (NSStoryboard *) mainStoryboard // 10.13
+{
+  return mainStoryboard;
+}
+
++ (instancetype) storyboardWithName: (NSStoryboardName)name
+                             bundle: (NSBundle *)bundle
+{
+  return AUTORELEASE([[NSStoryboard alloc] initWithName: name
+                                                 bundle: bundle]);
+}
+
+// Instance methods...
+- (void) dealloc
+{
+  RELEASE(_initialViewControllerId);
+  RELEASE(_applicationSceneId);
+  RELEASE(_scenesMap);
+  [super dealloc];
+}
+
+- (id) instantiateInitialController
+{
+  NSXMLDocument *xml = [_scenesMap objectForKey: _applicationSceneId];
+  NSData *xmlData = [xml XMLData];
+  GSModelLoader *loader = [GSModelLoaderFactory modelLoaderForFileType: @"xib"];
+  BOOL success = [loader loadModelData: xmlData
+                     externalNameTable: nil
+                              withZone: [self zone]];
+
+  if (success)
+    {
+      xml = [_scenesMap objectForKey: _applicationSceneId];
```

Comment:
Why do you have the identical code twice?

Reply:
This is a work in progress, one loads the application scene and the other loads the initial controller.  I am not sure if this should be split up or not.

Reply:
I have now changed these to be distinct.